opencypher/ast/query.py

```diff
@@ -61,6 +62,38 @@ def iter_parameters(self) -> Iterable[Parameter]:
                 yield from reading_clause.iter_parameters()
 
 
+@dataclass(frozen=True)
+class MultiPartQuery(Parameterized):
+    """
+    MultiPartQuery = { { ReadingClause, [SP] }, { UpdatingClause, [SP] }, With, [SP] }-, SinglePartQuery ;
+
+    """
+    updating_clauses: Sequence[UpdatingClause]
```

Comment:
I would change the order here so that `reading_clauses` comes before `updating_clauses` to match the grammar declaration order.

I would also try to allow default values where values are optional; this requires that you again change order so that required attributes come before optional ones.

```
from dataclasses import field
...

with_: With
single_part_query: "SinglePartQuery"
reading_clauses: Sequence[ReadingClause] = field(default_factory=list)
writing_clauses: Sequence[WritingClause] = field(default_factory=list)
```

HOWEVER, we're still not quite there because we're missing the ability to have multiple `With` declarations in a single multiple query. You actually want _another_ type for the `{ { ReadingClause, [SP] }, { UpdatingClause, [SP] }, With, [SP] }` grammar and to make `MultiPartQuery` have a `NonEmptySequence` of this new type, plus the `SinglePartQuery`.

I know...